STM: replace C++ low power ticker wrapper with a low level wrapper

LMESTM opened the pull request:
### Description
This series of commits aims at removing the latency limitation of the LP ticker wrapper. 

Current situation is that workaround from #9785 makes TICKLESS actually work without real deep sleep usage. #9785  was introduced because of latency issue with the C++ wrapper layer. 

With this Pull request the C+ wrapper layer is removed and replaced by a low level wrapper in lp ticker driver of STM targets. This direction was suggested by @c1728p9 and supported by @mprse 

Implementation of this low level wrapper require many specific implementation detailed in separate commits because of the way the LP TIMER HW is designed. 

In particular, the LP TIMER ticker cannot be programmed for a timestamp in the past, so we need to program a wake-up interrupt at the wrap-around (0xFFFF), then the next interrupt will be programmed again. Also the LP TIMER cannot cope with programming an interrupt within the next few low power clock 32KHz cycles, corresponding to the time it takes for CMPOK flag to be set, which means new comparator value is taken into account. Those design constraints are taken into account in this new implementation. 

The changes in this pull request have dependencies to a few test cases updates: 
#10536 #10698  #10700

Also changes to IDLE stack size are needed:
#10702  #10781 

Also this PR supersedes #10537 that will be closed. The commit from #10537 is actually included in this new PR. 

This PR should also solve #9962 and it reverts #10572 by enabling back LPTIM on NUCLEO_L073RZ.

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@c1728p9 @mprse @jeromecoutant 
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

The test results are good and show significant improvements on latencies. 
@mprse has verified that interrupt latencie is now lower than it was, allowing reliable serial communication @115200kbps on NUCLEO_L476RG target and 57600kbps on NUCLEO_L073RZ. 

Complete test suites have been run on 
TOOL_CHAIN=ARMC6,GCC_ARM,IAR
Executed targets : ['NUCLEO_WB55RG']
Executed targets : ['NUCLEO_L476RG']
Executed targets : ['NUCLEO_L073RZ'] 
And there is still a pending issue to be understood that might come from common ticker layer and was not seen before. @mprse is looking at it.

LMESTM committed:
LP ticker workaround

There is an errata in LPTIM specification that explains that CMP Flag
condition is not an exact match (COUNTER = MATCH) but rather a
comparison (COUNTER >= MATCH).

As a consequence the interrupt is firing early than expected when
programing a timestamp after the 0xFFFF wrap-around.

In order to
work-around this issue, we implement the below work-around.
In case timestamp is after the work-around, let's decide to program the
CMP value to 0xFFFF, which is the wrap-around value. There would anyway be
a wake-up at the time of wrap-around to let the OS update the system time.
When the wrap-around interrupt happen, OS will check the current time and
program again the timestamp to the proper value.

LMESTM committed:
Manage lp_ticker delay at low level

LP TICKER mbed-os wrapper needs to be disabled as it introduces too much latencies.

LP TICKER wrapper has been disabled and we need to managed the HW constraints at low level:
- main HW constraint is that once the comparator has been programmed once,
driver cannot program it again before CMPOK HW flag is set, which takes about 3 30us cycles.

To make it even more complex, the driver also needs to cope with "LP ticker workaround"

See commit:

LP ticker workaround

    There is an errata in LPTIM specification that explains that CMP Flag
    condition is not an exact match (COUNTER = MATCH) but rather a
    comparison (COUNTER >= MATCH).

Also the disable interrupt is more complete now:
- always check sleep manager status and restore it
- remove irq_handler as comparator is always programed and might get called
eventually when LP TICK is restarted
- reset delayed_prog

Also in set_interrupt, make sure interrupt does not fire early.
If needed, we decide to slightly delay the tick to cope with the HW limitation to
make sure it will fire as soon as HW is capable.

Functions are called under critical section as they may be called from
the IRQ handler now, not only from driver layer.

mprse commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -168,21 +182,32 @@ void lp_ticker_init(void)
+    __HAL_LPTIM_CLEAR_FLAG(&LptimHandle, LPTIM_FLAG_CMPOK);
+
+    /* Init is called with Interrupts disabled, so the CMPOK interrupt
+     * will not be handler. Let's mark it is now safe to write to LP counter */
```
typo: handler -> handled

LMESTM replied:
thx - fixed

mprse commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -217,42 +258,102 @@ uint32_t lp_ticker_read(void)
+            }
+        }
+        /* Then check if this target timestamp is not in the past, or close to wrap-around */
+        if((timestamp < last_read_counter) && (last_read_counter <= (0xFFFF - LP_TIMER_SAFE_GUARD))) {
```
We have already discussed this, but I suggest to add some comment about the crossing case (https://github.com/LMESTM/mbed/commit/17d2734fe4a41b582515906953af7da49f1cff64#r33648477).

This may be helpful in the future if some issues occur.

LMESTM replied:
indeed - fixed

mprse commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -217,42 +258,102 @@ uint32_t lp_ticker_read(void)
+        sleep_manager_unlock_deep_sleep();
+        sleep_manager_locked = false;
+    }
+    irq_handler = NULL;
```
Question: Why do we need `irq_handler` variable? 
I suggest to optimize the code and remove this variable.

LMESTM replied:
irq_handler is used to make sure that no user callback will be called after the calls to lp_ticker_disable_interrupt(). I agree this may not be needed, but I would maybe propose to do this optimization (and maybe others) in a future PR rather than this one.

LMESTM replied:
As other depending Prs are still under review, I added the few optimization changes as well

LMESTM committed:
STM32 targets: remove tickless from us ticker and delay ticker

For L0/L4/H7/F7/WB targets that have tickless enabled, remove the tickless from
us ticker and the delay ticks as the C++ wrapper layer is being removed
and replaced by the low layer handling.

For now, the few F4 targets with LPTIM are left with previous configuration
as test results are showing a few instabilities not yet understood.

LMESTM committed:
Update warning now that C++ wrapper is not needed anymore

Now we'd rather not use this wrapper and use instead the low level
wrapper implemented in this driver.

mprse reviewed (approved):
LGTM

I added only a few minor suggestions.

LMESTM committed:
Update comments and fix typo

LMESTM committed:
STM32: lp_ticker optim: remove irq_handler variable

LMESTM committed:
STM32: lp_ticker optim: initialize handle once only

mprse reviewed (approved):
Looks good

LMESTM committed:
STM32 with low level ticker wrapper require Extra Idle stack
